src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -348,57 +356,52 @@ impl<'a, 'tcx, 'rcx> DocFolder for LinkCollector<'a, 'tcx, 'rcx> {
-                            {
-                                let (value_kind, value_disambig)
-                                    = value_ns_kind(value_def.0, path_str)
-                                        .expect("struct and mod cases should have been \
-                                                 caught in previous branch");
-                                ambiguity_error(cx, &item.attrs, path_str,
-                                                "a", value_kind, &value_disambig,
-                                                "a", "macro", &format!("macro@{}", path_str));
-                            }
-                            (macro_def, None)
-                        } else if let Ok(type_def) =
+                            candidates.push(((macro_def, None), Namespace::Macro));
+                        }
+
+                        if let Ok(type_def) =
                             self.resolve(path_str, false, &current_item, parent_node)
                         {
-                            // It is imperative we search for not-a-value first
-                            // Otherwise we will find struct ctors for when we are looking
-                            // for structs, and the link won't work if there is something in
-                            // both namespaces.
-                            if let Ok(value_def) =
-                                self.resolve(path_str, true, &current_item, parent_node)
-                            {
-                                let kind = value_ns_kind(value_def.0, path_str);
-                                if let Some((value_kind, value_disambig)) = kind {
-                                    let (type_kind, article, type_disambig)
-                                        = type_ns_kind(type_def.0, path_str);
-                                    ambiguity_error(cx, &item.attrs, path_str,
-                                                    article, type_kind, &type_disambig,
-                                                    "a", value_kind, &value_disambig);
-                                    continue;
-                                }
-                            }
-                            type_def
-                        } else if let Ok(value_def) =
+                            candidates.push((type_def, Namespace::Type));
+                        }
+
+                        if let Ok(value_def) =
                             self.resolve(path_str, true, &current_item, parent_node)
                         {
-                            value_def
-                        } else {
+                            // Structs, variants, and mods exist in both namespaces, skip them.
+                            match value_def.0 {
+                                Def::StructCtor(..)
+                                | Def::Mod(..)
+                                | Def::Variant(..)
```

Comment:
`Mod` and `Variant` never live in the value namespace, so they can be removed from this list.